pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java

```diff
@@ -363,24 +343,27 @@ private void reapDeadBrokerPreallocations(Set<String> aliveBrokers) {
     @Override
     public Set<String> getAvailableBrokers() {
         try {
-            return availableActiveBrokers.get();
+            return new TreeSet<>(brokersData.listLocks(LoadManager.LOADBALANCE_BROKERS_ROOT).join());
         } catch (Exception e) {
             log.warn("Error when trying to get active brokers", e);
             return loadData.getBrokerData().keySet();
         }
     }
 
-    // Attempt to local the data for the given bundle in ZooKeeper.
+    // Attempt to local the data for the given bundle in metadata store
     // If it cannot be found, return the default bundle data.
     private BundleData getBundleDataOrDefault(final String bundle) {
         BundleData bundleData = null;
         try {
-            final String bundleZPath = getBundleDataZooKeeperPath(bundle);
-            final String quotaZPath = String.format("%s/%s", RESOURCE_QUOTA_ZPATH, bundle);
-            if (zkClient.exists(bundleZPath, null) != null) {
-                bundleData = readJson(zkClient.getData(bundleZPath, null, null), BundleData.class);
-            } else if (zkClient.exists(quotaZPath, null) != null) {
-                final ResourceQuota quota = readJson(zkClient.getData(quotaZPath, null, null), ResourceQuota.class);
+            Optional<BundleData> optBundleData = bundlesCache.get(getBundleDataPath(bundle)).join();
```

Comment:
can we use time bounded `get(timeout)` instead join to break the deadlock ?

Reply:
Rather than putting `get(Timeout)`, I prefer having timeouts on the async operations in the metadata store implementation.

Reply:
umm.. you mean it will be  metadata store-impl's responsibility to complete the returned CompletableFuture in certain time? ZK-Impl depends on ZK-Client and returns future. So, if we put timeout responsibility to impl class then Impl has to add extra handling to return time based future. this doesn't seem feasible solution.

Reply:
Yes, we can add the timeouts in the `AbstractMetadataStoreImpl`. Otherwise we only have timeouts for sync calls, but in most places we're doing (and we should do more) the async calls.